### PR TITLE
refactor: don't depend on multihash features

### LINKED ISSUE #3276: Don't depend on the multihash identity feature
This feature is broken and will be removed eventually. To avoid deprecation warnings for our users that they can't fix, let's remove it here first.

See:

- https://github.com/multiformats/rust-multihash/pull/260
- https://github.com/multiformats/rust-multihash/issues/194

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

All we need from the multihash is for it to be a data structure that we pass around. We only ever use the identity "hasher" and the sha256 hasher. Those are easily implemented without depending the (fairly heavy) machinery in `multihash`.

Unfortunately, this patch by itself does not yet lighten our dependency tree because `multiaddr` activates those features unconditionally. I opened a companion PR for this: https://github.com/multiformats/rust-multiaddr/pull/77.

https://github.com/multiformats/rust-multiaddr/pull/77 is another breaking change and we are trying to delay those at the moment. However, it will (hopefully) land eventually which should then be much easier to implement.

Fixes #3276.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: core/Cargo.toml [hunk 1]
```diff
@@ -19,7 +19,7 @@ instant = "0.1.11"
 libp2p-identity = { version = "0.1", path = "../identity", features = ["peerid", "ed25519"] }
 log = "0.4"
 multiaddr = { version = "0.17.0" }
-multihash = { version = "0.17.0", default-features = false, features = ["std", "multihash-impl", "identity", "sha2"] }
+multihash = { version = "0.17.0", default-features = false, features = ["std"] }
```
Comment (thomaseizinger) [inline core/Cargo.toml:22]: We re-export `multihash` so technically, this is a breaking change. I am tempted to release it as non-breaking as I'd argue that users should enable features for re-exported dependencies themselves.

**Comment by thomaseizinger** [comment]
Friendly ping @mxinden.

**Comment by thomaseizinger** [comment]
According to the Rust API evolution RFC, removing features from a dependency is not a breaking change because each crate should activate all features they depend on themselves. See https://rust-lang.github.io/rfcs/1105-api-evolution.html#minor-change-altering-the-use-of-cargo-features.

File: identity/src/peer_id.rs [hunk 2]
```diff
@@ -19,18 +19,24 @@
...
 
 use multiaddr::{Multiaddr, Protocol};
-use multihash::{Code, Error, Multihash};
+use multihash::{Code, Error, MultihashGeneric};
 use rand::Rng;
+use sha2::Digest as _;
 use std::{convert::TryFrom, fmt, str::FromStr};
 use thiserror::Error;
 
+type Multihash = MultihashGeneric<64>;
```
Comment (mxinden) [inline identity/src/peer_id.rs:33]: Can you include the rational for `64` once more?

File: protocols/kad/src/behaviour/test.rs [hunk 3]
```diff
@@ -138,7 +138,7 @@ fn build_fully_connected_nodes_with_config(
 }
 
 fn random_multihash() -> Multihash {
-    Multihash::wrap(Code::Sha2_256.into(), &thread_rng().gen::<[u8; 32]>()).unwrap()
+    Multihash::wrap(0x12, &thread_rng().gen::<[u8; 32]>()).unwrap()
```
Comment (mxinden) [inline protocols/kad/src/behaviour/test.rs]: I don't like us spreading magic numbers across the code-base. Can you at least document what `0x12` is?

File: protocols/kad/src/behaviour/test.rs [hunk 4]
```diff
@@ -1100,7 +1100,10 @@ fn disjoint_query_does_not_finish_before_all_paths_did() {
     let mut trudy = build_node(); // Trudy the intrudor, an adversary.
     let mut bob = build_node();
 
-    let key = Key::from(Code::Sha2_256.digest(&thread_rng().gen::<[u8; 32]>()));
+    let key = Key::from(
+        Multihash::wrap(0x12, &thread_rng().gen::<[u8; 32]>())
```
Comment (mxinden) [inline protocols/kad/src/behaviour/test.rs]: Same here.

File: protocols/kad/src/query/peers/closest.rs [hunk 5]
```diff
@@ -484,10 +484,8 @@ mod tests {
     fn random_peers<R: Rng>(n: usize, g: &mut R) -> Vec<PeerId> {
         (0..n)
             .map(|_| {
-                PeerId::from_multihash(
-                    Multihash::wrap(Code::Sha2_256.into(), &g.gen::<[u8; 32]>()).unwrap(),
-                )
-                .unwrap()
+                PeerId::from_multihash(Multihash::wrap(0x12, &g.gen::<[u8; 32]>()).unwrap())
```
Comment (mxinden) [inline protocols/kad/src/query/peers/closest.rs]: Same here.

File: protocols/kad/src/query/peers/closest.rs [hunk 6]
```diff
@@ -504,9 +502,7 @@ mod tests {
     impl Arbitrary for ArbitraryPeerId {
         fn arbitrary(g: &mut Gen) -> ArbitraryPeerId {
             let hash: [u8; 32] = core::array::from_fn(|_| u8::arbitrary(g));
-            let peer_id =
-                PeerId::from_multihash(Multihash::wrap(Code::Sha2_256.into(), &hash).unwrap())
-                    .unwrap();
+            let peer_id = PeerId::from_multihash(Multihash::wrap(0x12, &hash).unwrap()).unwrap();
```
Comment (mxinden) [inline protocols/kad/src/query/peers/closest.rs]: Same here.

File: protocols/kad/src/query/peers/closest/disjoint.rs [hunk 7]
```diff
@@ -530,9 +530,7 @@ mod tests {
     impl Arbitrary for ArbitraryPeerId {
         fn arbitrary(g: &mut Gen) -> ArbitraryPeerId {
             let hash: [u8; 32] = core::array::from_fn(|_| u8::arbitrary(g));
-            let peer_id =
-                PeerId::from_multihash(Multihash::wrap(Code::Sha2_256.into(), &hash).unwrap())
-                    .unwrap();
+            let peer_id = PeerId::from_multihash(Multihash::wrap(0x12, &hash).unwrap()).unwrap();
```
Comment (mxinden) [inline protocols/kad/src/query/peers/closest/disjoint.rs]: Same here.

File: protocols/kad/src/record.rs [hunk 8]
```diff
@@ -161,14 +161,13 @@ impl ProviderRecord {
...
     use super::*;
-    use libp2p_core::multihash::Code;
     use quickcheck::*;
     use std::time::Duration;
 
     impl Arbitrary for Key {
         fn arbitrary(g: &mut Gen) -> Key {
             let hash: [u8; 32] = core::array::from_fn(|_| u8::arbitrary(g));
-            Key::from(Multihash::wrap(Code::Sha2_256.into(), &hash).unwrap())
+            Key::from(Multihash::wrap(0x12, &hash).unwrap())
```
Comment (mxinden) [inline protocols/kad/src/record.rs]: And here.

File: protocols/kad/src/record/store/memory.rs [hunk 9]
```diff
@@ -216,12 +216,12 @@ impl RecordStore for MemoryStore {
...
 mod tests {
     use super::*;
-    use libp2p_core::multihash::{Code, Multihash};
+    use libp2p_core::multihash::Multihash;
     use quickcheck::*;
     use rand::Rng;
 
     fn random_multihash() -> Multihash {
-        Multihash::wrap(Code::Sha2_256.into(), &rand::thread_rng().gen::<[u8; 32]>()).unwrap()
+        Multihash::wrap(0x12, &rand::thread_rng().gen::<[u8; 32]>()).unwrap()
```
Comment (mxinden) [inline protocols/kad/src/record/store/memory.rs]: And this one.

**Review by mxinden — COMMENTED** [review]
Overall I am OK with this change. I understand the reasoning. I am a bit afraid of making hard-to-debug errors anywhere here.

File: protocols/kad/src/behaviour/test.rs (hunk 3 above)
Comment (thomaseizinger) [reply protocols/kad/src/behaviour/test.rs]: Sure, it is all just in tests so it doesn't really matter what we put there AFAIK.

**Comment by thomaseizinger** [comment]
@mxinden Please take another look.

**Review by mxinden — APPROVED** [review]
Thanks for the follow-ups. Looks good to me.

**Pull request merged**